Update multiple coordinators (#61)
hermeznetwork/docs: docs/developers/coordinator.md

```

### Connect Prover to Coordinator Node
Once you have verified the prover is working, you can connect it to the Hermez Coordinator by configuring the `cfg.sandbox.boot-coordinator.toml` configuration file.
You need to substitute sections `ServerProofs` with the updated URL where prover is deployed, and the `Circuit` section where the verifier smart contract is specified.

```
[[Coordinator.ServerProofs]]
#TODO: Add Prover URL
#URL = "http://localhost:9080"
```

```
[Coordinator.Circuit]
MaxTx = 400
NLevels = 32
```

At this point, you can stop the mock server if it is still running, and re-launch the coordinator as we saw in the previous section. The new prover will be running at http://localhost:9080 (or at the configured URL), and the two endpoints are `/status` and `/input`

## Launching a Synchronizer Node

> In synchronizer mode, the node is capable of keeping track of the rollup and consensus smart contracts, storing all the history of events and keeping the rollup

Reviewer: ```suggestion
In synchronizer mode, the node is capable of keeping track of the rollup and consensus smart contracts, storing all the history of events, and keeping the rollup
```

For this part of the tutorial, we are going to deploy the syncrhonizer node in testnet on Rinkeby.

1. Stop Coordinator node launched in localhost in previous steps.

  Stop prover, coordinator node and containers from previous phases as you will be working in testnet with a real Boot Coordinator node.
```shell
docker-compose -f docker-compose.sandbox.yaml down
```

2. Launch PostgreSQL database.

  The Hermez node in synchronizer mode needs to run on a separate database
```shell
docker run --rm --name hermez-db -p 5432:5432 -e POSTGRES_DB=hermez -e POSTGRES_USER=hermez -e POSTGRES_PASSWORD="yourpasswordhere" -d postgres
```

3. Start an Ethereum node in Rinkeby

You will need to run your own Ethreum node on Rinkeby. We recommend using Geth.
- Pre-built binaries for all platforms on our downloads page (https://geth.ethereum.org/downloads/).
- Ubuntu packages in our Launchpad PPA repository (https://launchpad.net/~ethereum/+archive/ubuntu/ethereum).
- OSX packages in our Homebrew Tap repository (https://github.com/ethereum/homebrew-ethereum).

> Sync this node with Rinkeby testnet where all Hermez's smart contracts are deployed. 

Reviewer: ```suggestion
Sync this node with Rinkeby testnet where all of Hermez's smart contracts are deployed. 
```

- **Web3** URL of your Rinkeby Ethereum node
- **SmartContracts** Double check that the address provided in the configuration file corresponds to the current Hermez Network contract deployed in Rinkeby

6. Launch `hermez-node` in synchronizer mode
```shell
./bin/heznode run --mode sync --cfg cmd/heznode/cfg.testnet.sync.toml
```

Once the Hermez node is launched, the API can be queried at the location specified in the configuration file in `API.Address` section, as well as at https://api.testnet.hermez.io/v1/ serviced by the Boot Coordinator node.

## Launching a Second Coordinator Node
In this part of the tutorial we will start a second Coordinator Node in testnet that will bid for the right to forge batches.

### Dependencies
- node 14+

### Start Coordinator in Testnet
1. Stop Synchronizer node and PostgreSQL container launched in previous steps.

2. Launch PostgreSQL database

```shell
docker run --rm --name hermez-db -p 5432:5432 -e POSTGRES_DB=hermez -e POSTGRES_USER=hermez -e POSTGRES_PASSWORD="yourpasswordhere" -d postgres
```
3. Launch Prover as shown [here](#launching-a-proof-server)

> 4. Create two Ethereum accounts in Rinkeby using Metamask wallet. One account is `forger` account (needs to pay for gas to forge batches in Ethereum and for bids in auction in HEZ), and the second is the `fee` account (receives the HEZ fees). The fees are collected in L2. You can convert from ETH to HEZ in [Uniswap](https://app.uniswap.org/#/swap?use=V2)

Reviewer: ```suggestion
4. Create two Ethereum accounts in Rinkeby using Metamask wallet. One account is the `forger` account (needs to pay for gas to forge batches in Ethereum and for bids in auction in HEZ), and the second is the `fee` account (receives the HEZ fees). The fees are collected in L2. You can convert from ETH to HEZ in [Uniswap](https://app.uniswap.org/#/swap?use=V2)
```

5. Create a Wallet with `fee` account Ethereum Private Key. 

> This wallet is needed to generate a Baby JubJub address where fees will be collected. There is an example code in the [SDK](https://github.com/hermeznetwork/hermezjs/blob/main/examples/create-wallet.js) that can be used. Simply substitue `EXAMPLES_WEB3_URL` by your Rinkeby Node URL and `EXAMPLES_PRIVATE_KEY1` by `fee` account private key.

Reviewer: ```suggestion
This wallet is needed to generate a Baby JubJub address where fees will be collected. There is an example code in the [SDK](https://github.com/hermeznetwork/hermezjs/blob/main/examples/create-wallet.js) that can be used. Simply substitute `EXAMPLES_WEB3_URL` by your Rinkeby Node URL and `EXAMPLES_PRIVATE_KEY1` by `fee` account private key.
```

7. Import the `forger` and `fee` Ethereum private keys into the keystore. 
```shell
./bin/heznode importkey --mode coord --cfg cmd/heznode/cfg.coord.toml --privatekey <FORGER ACCOUNT_PRIVATE KEY>
./bin/heznode importkey --mode coord --cfg cmd/heznode/cfg.coord.toml --privatekey <FEE_ACCOUNT PRIVATE KEY>
```
This private key corresponds to the new Coordinator node 

8. Launch New Coordinator Node
```shell
./bin/heznode run --mode coord --cfg cmd/heznode/cfg.testnet.coord.toml
```
The node will start synchronizing with the Hermez Network in testnet. This may take a while.

### Bidding Process
Once the node is synchronized, you can start bidding for the right to forge a batch.

1. Install cli-bidding

  `cli-bidding` is a tool that allows to register a Coordinator in Hermez Network and place bids in the auction.
```shell
git clone https://github.com/hermeznetwork/cli-bidding.git
```
Once downloaded, follow the installation steps in the [README](https://github.com/hermeznetwork/cli-bidding/blob/master/readme.md).
>NOTE that `PRIVATE_KEY_CLI_BIDDING` corresponds to the `forger` private key.

2. Approve HEZ transfers.

>   Before the coordinator can start bidding, it needs to approve the use of HEZ tokens. To do this go to [HEZ address in Etherscan](https://rinkeby.etherscan.io/token/0x2521bc90b4f5fb9a8d61278197e5ff5cdbc4fbf2), select `Contract` -> `Write Contract` -> `Approve` and set `spender address` to Coordinator address and `value` to quantity you want to approve. Recommendation is to set this quantity value very high.

Reviewer: ```suggestion
  Before the coordinator can start bidding, it needs to approve the use of HEZ tokens. To do this go to [HEZ address in Etherscan](https://rinkeby.etherscan.io/token/0x2521bc90b4f5fb9a8d61278197e5ff5cdbc4fbf2), select `Contract` -> `Write Contract` -> `Approve` and set `spender address` to Coordinator address and `value` to quantity you want to approve. The recommendation is to set this quantity value very high.
```

# Hermez Node

> This tutorial describes how to launch a Hermez node. It starts by explaining how to launch a Boot Coordinator in localhost.

Reviewer 2: ```suggestion
This tutorial describes how to launch an Hermez node. It starts by explaining how to launch a Boot Coordinator in localhost.
```

Author: It's correct as it is.

> It follows by describing how to initialize a Proof Server and how to connect it to the Boot Coordinator.

Reviewer: ```suggestion
Next, it describes how to initialize a Proof Server and how to connect it to the Boot Coordinator.
```

The next section describes how to spin up a second Hermez node in synchronizer mode to track the rollup status independently from the Boot Coordinator. This second node will be launched in Rinkeby testnet.
The last part of the tutorial includes an explanation on how to add a second Coordinator node to Hermez testnet that bids for the right to forge batches.
 
1. [Preparing the Environment](#preparing-the-environment) 
2. [Launching the Boot Coordinator](#launching-the-boot-coordinator)
3. [Launching a Proof Server](#launching-a-proof-server)
4. [Launching a Synchronizer Node](#launching-a-synchronizer-node)
5. [Launching a Second Coordinator](#launching-a-second-coordinator-node)

## Preparing the Environment

> Hermez node requires a PostgreSQL database and connectivity to an Ethereum node. In this part we describe how you can set this environment

Reviewer: ```suggestion
Hermez node requires a PostgreSQL database and connectivity to an Ethereum node. In this part, we describe how you can set this environment
```

up using docker containers.

> ### Dependencies

Reviewer 2: It could be useful to add the following links:
- install golang: https://golang.org/doc/install
- install docker: https://docs.docker.com/engine/install/ubuntu/
- install docker-compose: https://docs.docker.com/compose/install/

```yaml
version: "3.3"
services:
  hermez-db-test:
    image: postgres
    ports:
      - "5432:5432"
    environment:
      POSTGRES_USER: "hermez"
      POSTGRES_PASSWORD: "yourpasswordhere"
  privatebc:
    image:  public.ecr.aws/r7d5k1t8/hermez-geth:latest
    ports:
      - "8545:8545"
    environment:
      - DEV_PERIOD
    entrypoint: ["geth", "--http", "--http.addr", "0.0.0.0","--http.corsdomain", "*", "--http.vhosts" ,"*", "--ws", 
    "--ws.origins", "*", "--ws.addr", "0.0.0.0", "--dev", "--datadir", "/geth_data$DEV_PERIOD"]
```

Login to AWS public ECR to be able to download the Geth docker image:

```shell
export AWS_REGION=eu-west-3
aws ecr get-login-password --region eu-west-3 | docker login --username AWS --password-stdin 811278125247.dkr.ecr.eu-west-3.amazonaws.com
```

Start database and Geth node:
```shell
DEV_PERIOD=3 docker-compose -f docker-compose.sandbox.yaml up -d
```
This command will start a Geth node mining a block every 3 seconds. 
Database is available at port 5432. Geth node is available at port 8545.

> To stop containers

Reviewer 2: ```suggestion
To stop containers:
```

 "chainId" : 1337
}
```

4. Configure `cfg.buidler.toml` file. The configuration file located at `cli/node/cfg.buidler.toml` needs to be customized to be able to deploy the Hermez node using the predeployed contracts.

You will need to change the `PriceUpdater` section to limit the frequency of the updates so that the logs are not clogged with price update information, the `SmartContract` section to 
reflect the location of the deployed contracts, and the address of the Coordinator node. Just substitute the sections below in the `cfg.buidler.toml` file.

```
[PriceUpdater]
Interval = "1000s"
```

```
[SmartContracts]
Rollup   = "0x10465b16615ae36F350268eb951d7B0187141D3B"
Auction  = "0x317113D2593e3efF1FfAE0ba2fF7A61861Df7ae5"
WDelayer = "0x8EEaea23686c319133a7cC110b840d1591d9AeE0"
TokenHEZ = "0x5E0816F0f8bC560cB2B9e9C87187BeCac8c2021F"
TokenHEZName = "Hermez Network Token"
```

```
[Coordinator]
ForgerAddress = "0xDcC5dD922fb1D0fd0c450a0636a8cE827521f0eD" # Non-Boot Coordinator
```
For more information on the parameters in the configuration file, see [this](https://github.com/hermeznetwork/hermez-node/blob/master/config/config.go#L57).

## Launching the Boot Coordinator

> 1. Copy `cfg.buidler.toml` file. Configuration file can be found in `hermez-node/cli/node` folder

Reviewer: ```suggestion
1. Copy `cfg.buidler.toml` file. The configuration file can be found in `hermez-node/cli/node` folder
```

```shell
cp cfg.buidler.toml cfg.boot-coordinator.cfg
```

> 1. Import the Coordinator Ethereum private key into the keystore. 

Reviewer 2: ```suggestion
2. Import the Coordinator Ethereum private key into the keystore. 
```

and the following indexes

go build .
./cli -d 15s -a 0.0.0.0:3000
```
The `hermez-node` repository provides a mock proof server that generates mock proofs every 15 seconds. The mock prover is launched at http://localhost:3000, and it exports two endpoints:
- GET /api/status: Queries the prover's status.
- POST /api/input: Starts the generation of a new proof.

3. Wipe SQL database

Before starting the Coordinator node, you may want to wipe the pre-existing SQL database.
```shell
./node wipesql --mode coord --cfg cfg.boot-coordinator.toml 
```

4. Launch `hermez-node`
```shell
./node run --mode coord --cfg cfg.boot-coordinator.toml
```

Once the Hermez node is launched, the API can be queried at `localhost:8086/v1`.

For more information, check the [README](https://github.com/hermeznetwork/hermez-node/tree/master/cli/node) file.

## Launching a Proof Server
We will use [rapidsnark](https://github.com/iden3/rapidsnark) as the Hermez proof server. `rapidsnarks` is a zkSnark proof generator written in C++.
It is recommended to run the proof server in servers with 32+ cores and 64 GB+ of RAM (AWS c5a.8xlarge or equivalent).

> rapidsnark requires a host CPU that supports ADX extensions. 

### Dependencies

> - node v12+

Reviewer 2: It could be useful to add link to node install: https://nodejs.org/en/download/

```
3. Launch prover
```shell
cd ..
./rapidsnark/build/proverServer circuit-344-32-256-64.dat circuit-344-32-256-64_hez3_final.zkey
```
Prover is deployed at port 9080.

4. Check prover status
```shell
curl -i -H "Accept: application/json" -H "Content-Type: application/json" -X GET http://localhost:9080/status
```

### Generate a Prover Input File

1. Clone [circuits](https://github.com/hermeznetwork/circuits.git) repository
```shell
git clone https://github.com/hermeznetwork/circuits.git
```

2. Install dependencies
```shell
cd circuits
npm install
cd tools
```

In this example we are working with `circuit-344-32-236-64_hez1.zkey`, which corresponds to a circuit with 344 transactions, 32 levels, 256 maxL1Tx and 64 maxFeeTx.

3. Generate Input file

To generate a new input file with empty transactions:

```shell
node build-circuit.js input 344 32 256 64
```
This command generates a new input file `rollup-344-32-256-64/input-344-32-256-64.json`

To generate a new input file with random transactions
```shell
node generate-input.js 256 88 344 32 256 64
``` 
This will create a new input file called `inputs-256.json` 

### Generate a New Proof

> You can use curl to post any of the inputs generated in the previous step.

Reviewer 2: ```suggestion
You can use `curl` to post any of the inputs generated in the previous step.
```

## Launching a Synchronizer Node
In synchronizer mode, the node is capable of keeping track of the rollup and consensus smart contracts, storing all the history of events and keeping the rollup
state updated, handling reorgs when they happen. This mode is intended for entities that want to gather all the rollup data by themselves and not rely on third party APIs.
For this part of the tutorial, we are going to deploy the syncrhonizer node in testnet on Rinkeby.

1. Stop Coordinator node launched in localhost in previous steps.

Stop prover, coordinator node and containers from previous phases as you will be working in testnet with a real Boot Coordinator node.
```shell
docker-compose -f docker-compose.sandbox.yaml down
```

2. Launch PostgreSQL database

The Hermez node in synchronizer mode needs to run a separate database
```shell
docker run --rm --name hermez-db -p 5432:5432 -e POSTGRES_DB=hermez -e POSTGRES_USER=hermez -e POSTGRES_PASSWORD="yourpasswordhere" -d postgres
```

3. Start an Ethereum node in Rinkeby

You can use Infura or you can run your own node using Geth.

> - Pre-built binaries for all platforms on our downloads page (https://geth.ethereum.org/downloads/).

Reviewer 2: ```suggestion
- Pre-built binaries for all platforms on downloads page (https://geth.ethereum.org/downloads/).
```

> - Ubuntu packages in our Launchpad PPA repository (https://launchpad.net/~ethereum/+archive/ubuntu/ethereum).

Reviewer 2: ```suggestion
- Ubuntu packages in Launchpad PPA repository (https://launchpad.net/~ethereum/+archive/ubuntu/ethereum).
```

> - OSX packages in our Homebrew Tap repository (https://github.com/ethereum/homebrew-ethereum).

Reviewer 2: ```suggestion
- OSX packages in Homebrew Tap repository (https://github.com/ethereum/homebrew-ethereum).
```

> Sync this node with Rinkeby testnet where all Hermez's smart contract are deployed. 

Reviewer: ```suggestion
Sync this node with Rinkeby testnet where all Hermez's smart contracts are deployed. 
```

4. Get contract addresses

Query Testnet API for the addresses of the Hermez smart contracts.

```shell
curl -i -H "Accept: application/json" -H "Content-Type: application/json" -X GET api.testnet.hermez.io/v1/config
```

> At this moment, the contracts are deployed in these addresses

Reviewer 2: ```suggestion
At this moment, the contracts are deployed in these addresses:
```

TokenHEZName = "Hermez Network Token"

[RecommendedFeePolicy]
# Strategy used to calculate the recommended fee that the API will expose.
# Available options:
# - Static: always return the same value (StaticValue) in USD
# - AvgLastHour: calculate using the average fee of the forged transactions during the last hour
PolicyType = "Static"
StaticValue = 0.99
```

6. Launch `hermez-node` in synchronizer mode
```shell
./node run --mode sync --cfg cfg.sync.toml
```

Once the Hermez node is launched, the API can be queried at `localhost:8086/v1` (as well as at https://api.testnet.hermez.io/v1/ serviced by the Boot Coordinator node).

## Launching a Second Coordinator Node
In this part of the tutorial we will start a second Coordinator Node in testnet that will bid for the right to forge batches.

### Dependencies
- node 12+

### Start Coordinator in Testnet
1. Stop Synchronizer node and PostgreSQL container launched in previous steps.

2. Launch PostgreSQL database

```shell
docker run --rm --name hermez-db -p 5432:5432 -e POSTGRES_DB=hermez -e POSTGRES_USER=hermez -e POSTGRES_PASSWORD="yourpasswordhere" -d postgres
```
3. Launch Prover as shown [here](#launching-a-proof-server)

> 4. Create two Ethereum accounts in Rinkeby using Metamask wallet. One account is `forger` account (needs to pay to forge batches), and the second is the `fee` account (receives the fees). 

Reviewer 2: Maybe it is worth it to mention that the fees are collected on L2

5. Create a Wallet with `fee` account Ethereum Private Key. 

> This wallet is needed to generate a Baby JubJub address where fees will be collected. There is an example code in the [SDK](https://github.com/hermeznetwork/hermezjs/blob/main/examples/create-wallet.js) that can be used. Simply substitue `EXAMPLES_WEB3_URL` by your Rinkeby Node URL and EXAMPLES_PRIVATE_KEY1` by `fee` account private key.

Reviewer 2: ```suggestion
This wallet is needed to generate a Baby JubJub address where fees will be collected. There is an example code in the [SDK](https://github.com/hermeznetwork/hermezjs/blob/main/examples/create-wallet.js) that can be used. Simply substitue `EXAMPLES_WEB3_URL` by your Rinkeby Node URL and `EXAMPLES_PRIVATE_KEY1` by `fee` account private key.
```

NLevels = 32
```

Adjust Gas
```
[Coordinator.EthClient.ForgeBatchGasCost]
Fixed = 900000
```

7. Import the `forger` and `fee` Ethereum private keys into the keystore. 
```shell
./node importkey --mode coord --cfg cfg.coord.toml --privatekey <FORGER ACCOUNT_PRIVATE KEY>
./node importkey --mode coord --cfg cfg.coord.toml --privatekey <FEE_ACCOUNT PRIVATE KEY>
```
This private key corresponds to the new Coordinator node 

8. Get ETH and HEZ to the `forger` account.

You need to fund the `forger` account with some ETH and some HEZ. 
You can also use Metamask and generate a `send` transaction from `forger` account to the HEZ token address. This transaction will add 100HEZ to the `forger` account.

9. Launch New Coordinator Node
```shell
./node run --mode coord --cfg cfg.coord.toml
```
The node will start synchronizing with the Hermez Network in testnet. This may take a while.

### Bidding Process
Once the node is synchronized, you can start bidding for the right to forge a batch.

1. Install CLI-bidding

CLI-bidding is a tool that allows to register a Coordinator in Hermez Network and place bids in the auction.

```shell
git clone https://github.com/hermeznetwork/CLI-bidding.git
```

> Once downloaded, follow the installation steps. `PRIVATE_KEY_CLI_BIDDING` corresponds the `forger` private key.

Reviewer 2: ```suggestion
Once downloaded, follow the installation steps. `PRIVATE_KEY_CLI_BIDDING` corresponds to the `forger` private key.
```

2. Register Forger

> Using `CLI-bidding`, you need to register the new Coordinator API URL. In my case, we have the Coordinator node running at `http://134.255.190.114:8086`

Reviewer: ```suggestion
Using `CLI-bidding`, you need to register the new Coordinator API URL. In our case, we have the Coordinator node running at `http://134.255.190.114:8086`
```

3. Get Current Slot in Hermez bid

Take a look at the current slot being bid in Hermez. When bidding, you need to bid at least 2 slots after the curent slot
```shell
curl -i -H "Accept: application/json" -H "Content-Type: application/json" -X GET http://localhost:8086/v1/state
```
In the json returned, there is a field, `network.currentSlot`, with the current Hermez slot. Let's assume this `currentSlot` is 4200

4. Check Minimum bid

Auctions have a minimum bid. You can check the current minimum bid for the upcomming slots
```shell
node src/biddingCLI.js slotinfo --startingSlot 4200 --endingSlot 4210
```

In our case, minimum bidding is set to 1.1 HEZ.

4. Bidding Process

Send a simple bid of 1.1x10^18 HEZ for slot 4200. Parameter `amount` is the amount to be transferred to the auction smart contract. Parameter `bidAmount` is the actual bid amount. Thus `amount` >= `bidAmount` unless there are some existing funds previously transferred.

```shell
node src/biddingCLI.js bid --amount 1100000000000000000 --slot 4200 --bidAmount 1100000000000000000
```

> If the bidding process is sucessful, an Etherscan URL with the transaction id is returned to verify transaction.

Reviewer 2: ```suggestion
If the bidding process is successful, an Etherscan URL with the transaction id is returned to verify transaction.
```

You can check the allocated nextForgers using 
```shell
curl -i -H "Accept: application/json" -H "Content-Type: application/json" -X GET http://localhost:8086/v1/state
```

> When the time to forge acutioned slots comse, the node you supplied will be the one forging the upcoming batches.

Reviewer: ```suggestion
When the time to forge the auctioned slots comes, the node you supplied will be the one forging the upcoming batches.
```

> When the time to forge the acutioned slots comes, the node you supplied will be the one forging the upcoming batches.

Reviewer 2: ```suggestion
When the time to forge the auctioned slots comes, the node you supplied will be the one forging the upcoming batches.
```
